man/my_print_defaults.1
'\" t
.\"     Title: my_print_defaults
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/25/2026
.\"    Manual: MySQL Database System
.\"    Source: MySQL 9.7
.\"  Language: English
.\"
.TH "MY_PRINT_DEFAULTS" "1" "06/25/2026" "MySQL 9\&.7" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
my_print_defaults \- display options from option files
.SH "SYNOPSIS"
.HP \w'\fBmy_print_defaults\ [\fR\fB\fIoptions\fR\fR\fB]\ \fR\fB\fIoption_group\fR\fR\fB\ \&.\&.\&.\fR\ 'u
\fBmy_print_defaults [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIoption_group\fR\fR\fB \&.\&.\&.\fR
.SH "DESCRIPTION"
.PP
\fBmy_print_defaults\fR
displays the options that are present in option groups of option files\&. The output indicates what options are used by programs that read the specified option groups\&. For example, the
\fBmysqlcheck\fR
program reads the
[mysqlcheck]
and
[client]
option groups\&. To see what options are present in those groups in the standard option files, invoke
\fBmy_print_defaults\fR
like this:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBmy_print_defaults mysqlcheck client\fR
\-\-user=myusername
\-\-password=\fIpassword\fR
\-\-host=localhost
.fi
.if n \{\
.RE
.\}
.PP
The output consists of options, one per line, in the form that they would be specified on the command line\&.
.PP
\fBmy_print_defaults\fR
supports the following options\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-help\fR,
\fB\-?\fR
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-config\-file=\fR\fB\fIfile_name\fR\fR,
\fB\-\-defaults\-file=\fR\fB\fIfile_name\fR\fR,
\fB\-c \fR\fB\fIfile_name\fR\fR
Read only the given option file\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-debug=\fR\fB\fIdebug_options\fR\fR,
\fB\-# \fR\fB\fIdebug_options\fR\fR
Write a debugging log\&. A typical
\fIdebug_options\fR
string is
d:t:o,\fIfile_name\fR\&. The default is
d:t:o,/tmp/my_print_defaults\&.trace\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-extra\-file=\fR\fB\fIfile_name\fR\fR,
\fB\-\-extra\-file=\fR\fB\fIfile_name\fR\fR,
\fB\-e \fR\fB\fIfile_name\fR\fR
Read this option file after the global option file but (on Unix) before the user option file\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-group\-suffix=\fR\fB\fIsuffix\fR\fR,
\fB\-g \fR\fB\fIsuffix\fR\fR
In addition to the groups named on the command line, read groups that have the given suffix\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-login\-path=\fR\fB\fIname\fR\fR,
\fB\-l \fR\fB\fIname\fR\fR
Read options from the named login path in the
\&.mylogin\&.cnf
login path file\&. A
\(lqlogin path\(rq
is an option group containing options that specify which MySQL server to connect to and which account to authenticate as\&. To create or modify a login path file, use the
\fBmysql_config_editor\fR
utility\&. See
mysql_config_editor(1)\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-no\-login\-paths\fR
Skips reading options from the login path file\&.
.sp
See
\fB\-\-login\-path\fR
for related information\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-no\-defaults\fR,
\fB\-n\fR
Return an empty string\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-show\fR,
\fB\-s\fR
\fBmy_print_defaults\fR
masks passwords by default\&. Use this option to display passwords as cleartext\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-verbose\fR,
\fB\-v\fR
Verbose mode\&. Print more information about what the program does\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-version\fR,
\fB\-V\fR
Display version information and exit\&.
.RE
.SH "COPYRIGHT"
.br
.PP
Copyright \(co 1997, 2026, Oracle and/or its affiliates.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Oracle Corporation (http://dev.mysql.com/).


ChangeLog
2010-03-18	The InnoDB Team

	* CMakeLists.txt:
	Fix Bug#52102 InnoDB Plugin shows performance drop compared to
	InnoDB (Windows)

2010-03-18	The InnoDB Team

	* buf0buf.ic:
	When comparing the time of the first access to a block against
	innodb_old_blocks_time, use 32-bit arithmetics. The comparison was
	incorrect on 64-bit systems.

2010-03-11	The InnoDB Team

	* buf0buf.h, buf0buf.ic:
	Fix and clarify the latching of some buf_block_t members.
	Note that check_index_page_at_flush is not protected by any mutex.
	Note and assert that lock_hash_val is protected by the rw-latch.

2010-03-10	The InnoDB Team

	* trx/trx0sys.c:
	Fix Bug#51653 outdated reference to set-variable

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb_bug21704.result,
	mysql-test/innodb_bug47621.result, mysql-test/innodb_bug47621.test:
	Fix Bug#47621 MySQL and InnoDB data dictionaries will become out of
	sync when renaming columns

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#51356 Many Valgrind errors in error messages
	with concurrent DDL

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, handler/handler0alter.cc,
	mysql-test/innodb_bug51378.result, mysql-test/innodb_bug51378.test:
	Fix Bug#51378 Init 'ref_length' to correct value, in case an out
	of bound MySQL primary_key

2010-03-10	The InnoDB Team

	* log/log0recv.c:
	Remove a bogus assertion about page numbers exceeding 0x90000000
	in the redo log. Abort when encountering a corrupted redo log
	record, unless innodb_force_recovery is set.

2010-03-09	The InnoDB Team

	* handler/ha_innodb.cc:
	Make SHOW ENGINE INNODB MUTEX STATUS display SUM(os_waits)
	for the buffer pool block mutexes and locks.

2010-03-08	The InnoDB Team

	* fil/fil0fil.c:
	Fix ALTER TABLE ... IMPORT TABLESPACE of compressed tables.

2010-03-03	The InnoDB Team

	* handler/handler0alter.cc, innodb-index.result, innodb-index.test,
	innodb.result, innodb.test:
	Disallow a duplicate index name when creating an index.

2010-02-11	The InnoDB Team

	* include/mem0mem.h, include/mem0mem.ic, mem/mem0mem.c:
	Fix Bug#49535 Available memory check slows down crash
	recovery tens of times

2010-02-09	The InnoDB Team

	* buf/buf0buf.c:
	Fix Bug#38901 InnoDB logs error repeatedly when trying to load
	page into buffer pool

2010-02-09	The InnoDB Team

	* srv/srv0srv.c:
	Let the master thread sleep if the amount of work to be done is
	calibrated as taking less than a second.

2010-02-04	The InnoDB Team

	* btr/btr0btr.c, btr/btr0cur.c, btr/btr0pcur.c, buf/buf0buf.c,
	include/btr0btr.h, include/btr0cur.h, include/btr0pcur.h,
	include/btr0pcur.ic, include/buf0buf.h, row/row0ins.c, row/row0sel.c:
	Pass the file name and line number of the caller of the
	b-tree cursor functions to the buffer pool requests, in order
	to make the latch diagnostics more accurate.

2010-02-03	The InnoDB Team

	* lock/lock0lock.c:
	Fix Bug#49001 SHOW INNODB STATUS deadlock info incorrect
	when deadlock detection aborts

2010-02-03	The InnoDB Team

	* buf/buf0lru.c:
	Fix Bug#35077 Very slow DROP TABLE (ALTER TABLE, OPTIMIZE TABLE)
	on compressed tables

2010-02-03	The InnoDB Team

	* handler/ha_innodb.cc, include/row0mysql.h, row/row0mysql.c:
	Clean up CHECK TABLE error handling.

2010-02-01	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb-autoinc.test,
	mysql-test/innodb-autoinc.result,
	mysql-test/innodb-autoinc-44030.test,
	mysql-test/innodb-autoinc-44030.result:
	Fix Bug#49497 Error 1467 (ER_AUTOINC_READ_FAILED) on inserting
	a negative value

2010-01-27	The InnoDB Team

	* include/row0mysql.h, log/log0recv.c, row/row0mysql.c:
	Drop temporary tables at startup.
	This addresses the third aspect of
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.

2010-01-21	The InnoDB Team

	* buf/buf0buf.c:
	Do not merge buffered inserts to compressed pages before
	the redo log has been applied in crash recovery.

2010-01-13	The InnoDB Team

	* row/row0sel.c:
	On the READ UNCOMMITTED isolation level, do not attempt to access
	a clustered index record that has been marked for deletion. The
	built-in InnoDB in MySQL 5.1 and earlier would attempt to retrieve
	a previous version of the record in this case.

2010-01-13	The InnoDB Team

	* buf/buf0buf.c:
	When disabling the adaptive hash index, check the block state
	before checking block->is_hashed, because the latter may be
	uninitialized right after server startup.

2010-01-12	The InnoDB Team

	* handler/ha_innodb.cc, handler/ha_innodb.h:
	Fix Bug#46193 crash when accessing tables after enabling
	innodb_force_recovery option

2010-01-12	The InnoDB Team

	* row/row0mysql.c:
	Fix Bug#49238 Creating/Dropping a temporary table while at 1023
	transactions will cause assert.

2009-12-02	The InnoDB Team

	* srv/srv0start.c:
	Display the zlib version number at startup.
	InnoDB compressed tables use zlib, and the implementation depends
	on the zlib function compressBound(), whose definition was slightly
	changed in zlib version 1.2.3.1 in 2006. MySQL bundles zlib 1.2.3
	from 2005, but some installations use a more recent zlib.

2009-11-30	The InnoDB Team

	* dict/dict0crea.c, dict/dict0mem.c, dict/dict0load.c,
	dict/dict0boot.c, fil/fil0fil.c, handler/ha_innodb.cc,
	include/dict0mem.h, row/row0mysql.c:
	Fix the bogus warning messages for non-existing temporary
	tables that were reported in
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.
	The actual crash recovery bug was corrected on 2009-04-29.

2009-11-27	The InnoDB Team

	InnoDB Plugin 1.0.6 released

2009-11-20	The InnoDB Team

	* handler/ha_innodb.cc:
	Add a workaround to prevent a crash due to Bug#45961 DDL on
	partitioned innodb tables leaves data dictionary in an inconsistent
	state

2009-11-19	The InnoDB Team

	* btr/btr0btr.c:
	Fix Bug#48469 when innodb tablespace is configured too small, crash
	and corruption!

2009-11-19	The InnoDB Team

	* data/data0type.c:
	Fix Bug#48526 Data type for float and double is incorrectly reported
	in InnoDB table monitor


utilities/my_print_defaults.cc

/*
   Copyright (c) 2000, 2026, Oracle and/or its affiliates.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is designed to work with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have either included with
   the program or referenced in the documentation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

// Print all parameters in a default file that will be given to some program.

#include <sys/types.h>
#include <cstdio>
#include <cstdlib>

#include "m_string.h"
#include "my_alloc.h"
#include "my_compiler.h"
#include "my_dbug.h"
#include "my_default.h"
#include "my_getopt.h"
#include "my_sys.h"
#include "mysql/service_mysql_alloc.h"
#include "print_version.h"
#include "welcome_copyright_notice.h" /* ORACLE_WELCOME_COPYRIGHT_NOTICE */

const char *config_file = "my"; /* Default config file */
static char *my_login_path;
static bool *show_passwords;
uint verbose = 0, opt_defaults_file_used = 0;
const char *default_dbug_option = "d:t:o,/tmp/my_print_defaults.trace";

static struct my_option my_long_options[] = {
    /*
      NB: --config-file is troublesome, because get_defaults_options() doesn't
      know about it, but we pretend --config-file is like --defaults-file.  In
      fact they behave differently: see the comments at the top of
      mysys/default.c for how --defaults-file should behave.

      This --config-file option behaves as:
      - If it has a directory name part (absolute or relative), then only this
        file is read; no error is given if the file doesn't exist
      - If the file has no directory name part, the standard locations are
        searched for a file of this name (and standard filename extensions are
        added if the file has no extension)
    */
    {"config-file", 'c',
     "Deprecated, please use --defaults-file instead. "
     "Name of config file to read; if no extension is given, default "
     "extension (e.g., .ini or .cnf) will be added",
     &config_file, &config_file, nullptr, GET_STR, REQUIRED_ARG, 0, 0, 0,
     nullptr, 0, nullptr},
#ifdef NDEBUG
    {"debug", '#', "This is a non-debug version. Catch this and exit", nullptr,
     nullptr, nullptr, GET_DISABLED, OPT_ARG, 0, 0, 0, nullptr, 0, nullptr},
#else
    {"debug", '#', "Output debug log", &default_dbug_option,
     &default_dbug_option, nullptr, GET_STR, OPT_ARG, 0, 0, 0, nullptr, 0,
     nullptr},
#endif
    {"defaults-file", 'c',
     "Like --config-file, except: if first option, "
     "then read this file only, do not read global or per-user config "
     "files; should be the first option",
     &config_file, &config_file, nullptr, GET_STR, REQUIRED_ARG, 0, 0, 0,
     nullptr, 0, nullptr},
    {"defaults-extra-file", 'e',
     "Read this file after the global config file and before the config "
     "file in the users home directory; should be the first option",
     &my_defaults_extra_file, &my_defaults_extra_file, nullptr, GET_STR,
     REQUIRED_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"defaults-group-suffix", 'g',
     "In addition to the given groups, read also groups with this suffix",
     &my_defaults_group_suffix, &my_defaults_group_suffix, nullptr, GET_STR,
     REQUIRED_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"extra-file", 'e', "Deprecated. Synonym for --defaults-extra-file.",
     &my_defaults_extra_file, &my_defaults_extra_file, nullptr, GET_STR,
     REQUIRED_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"no-defaults", 'n',
     "Ignore reading of default option file(s), "
     "except for login file.",
     nullptr, nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0,
     nullptr},
    {"login-path", 'l', "Path to be read from under the login file.",
     &my_login_path, &my_login_path, nullptr, GET_STR, REQUIRED_ARG, 0, 0, 0,
     nullptr, 0, nullptr},
    {"no-login-paths", 'x', "Ignore reading of the login file.", nullptr,
     nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"show", 's', "Show passwords in plain text.", &show_passwords,
     &show_passwords, nullptr, GET_BOOL, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"help", '?', "Display this help message and exit.", nullptr, nullptr,
     nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"verbose", 'v', "Increase the output level", nullptr, nullptr, nullptr,
     GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"version", 'V', "Output version information and exit.", nullptr, nullptr,
     nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {nullptr, 0, nullptr, nullptr, nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0,
     0, nullptr, 0, nullptr}};

static void usage(bool version) {
  print_version();
  if (version) return;
  puts(ORACLE_WELCOME_COPYRIGHT_NOTICE("2000"));
  puts(
      "Prints all arguments that is give to some program using the default "
      "files");
  printf("Usage: %s [OPTIONS] groups\n", my_progname);
  my_print_help(my_long_options);
  my_print_default_files(config_file);
  my_print_variables(my_long_options);
  printf("\nExample usage:\n%s --defaults-file=example.cnf client mysql\n",
         my_progname);
}

static bool get_one_option(int optid,
                           const struct my_option *opt [[maybe_unused]],
                           char *argument [[maybe_unused]]) {
  switch (optid) {
    case 'c':
      opt_defaults_file_used = 1;
      break;
    case 'n':
      break;
    case 'I':
    case '?':
      usage(false);
      exit(0);
    case 'v':
      verbose++;
      break;
    case 'V':
      usage(true);
      exit(0);
    case '#':
      DBUG_PUSH(argument ? argument : default_dbug_option);
      break;
  }
  return false;
}

static int get_options(int *argc, char ***argv) {
  int ho_error;

  if ((ho_error = handle_options(argc, argv, my_long_options, get_one_option)))
    exit(ho_error);

  if (*argc < 1) {
    usage(false);
    return 1;
  }
  return 0;
}

int main(int argc, char **argv) {
  int count, error, args_used;
  const char **load_default_groups;
  char *tmp_arguments[6];
  char **argument, **arguments, **org_argv;
  char *defaults, *extra_defaults, *group_suffix, *login_path;
  bool found_no_login_paths_dummy = false;
  MY_INIT(argv[0]);

  org_argv = argv;
  args_used = get_defaults_options(argc, argv, &defaults, &extra_defaults,
                                   &group_suffix, &login_path, false,
                                   &found_no_login_paths_dummy);

  /* Copy defaults-xxx arguments & program name */
  count = args_used + 1;
  arguments = tmp_arguments;
  memcpy((char *)arguments, (char *)org_argv, count * sizeof(*org_argv));
  arguments[count] = nullptr;

  /* Check out the args */
  if (!(load_default_groups = static_cast<const char **>(my_malloc(
            PSI_NOT_INSTRUMENTED, (argc + 1) * sizeof(char *), MYF(MY_WME)))))
    exit(1);
  if (get_options(&argc, &argv)) exit(1);
  memcpy((char *)load_default_groups, (char *)argv, (argc + 1) * sizeof(*argv));

  MEM_ROOT alloc{PSI_NOT_INSTRUMENTED, 512};
  if ((error = load_defaults(config_file, load_default_groups, &count,
                             &arguments, &alloc))) {
    if (verbose && opt_defaults_file_used) {
      if (error == 1)
        fprintf(stderr, "WARNING: Defaults file '%s' not found!\n",
                config_file);
      /* This error is not available now. For the future */
      if (error == 2)
        fprintf(stderr, "WARNING: Defaults file '%s' is not a regular file!\n",
                config_file);
    }
    error = 2;
    exit(error);
  }

  for (argument = arguments + 1; *argument; argument++)
    if (!my_getopt_is_args_separator(*argument)) /* skip arguments separator */
    {
      if (!(show_passwords) && strncmp(*argument, "--password=", 11) == 0)
        puts("--password=*****");
      else
        puts(*argument);
    }
  my_free(load_default_groups);

  exit(0);
}
